File: CSharpFundamentals/CSharpOOPbasics/InheritanceEx/Mankind/Human.cs
using System;
using System.Text;

namespace Mankind
{
    public class Human
    {
        private const int FIRST_NAME_MIN_LENGTH = 4;
        private const int LAST_NAME_MIN_LENGTH = 3;
        private string firstName;
        private string lastName;

        public Human(string firstName, string lastName)
        {
            this.FirstName = firstName;
            this.LastName = lastName;
        }

        protected string LastName
        {
            get { return lastName; }
            set
            {
                if (!char.IsUpper(value[0]))
                {
                    throw new ArgumentException($"Expected upper case letter! Argument: {nameof(lastName)}");
                }
                if (value.Length < LAST_NAME_MIN_LENGTH)
                {
                    throw new ArgumentException($"Expected length at least {LAST_NAME_MIN_LENGTH} symbols! Argument: {nameof(lastName)}");
                }
                lastName = value;
            }
        }

        protected string FirstName
        {
            get { return firstName; }
            set
            {
                if (!char.IsUpper(value[0]))
                {
                    throw new ArgumentException($"Expected upper case letter! Argument: {nameof(firstName)}");
                }
                if (value.Length < FIRST_NAME_MIN_LENGTH)
                {
                    throw new ArgumentException($"Expected length at least {FIRST_NAME_MIN_LENGTH} symbols! Argument: {nameof(firstName)}");
                }
                firstName = value;
            }
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.AppendLine($"First Name: {this.FirstName}");
            sb.AppendLine($"Last Name: {this.LastName}");
            return sb.ToString().TrimEnd();
        }
    }
}

File: JavaDB/DatabaseAdvancedFrameworks/JavaOOPPrinciplesEx/Mankind/models/Human.java
package Mankind.models;

public abstract class Human {
    private static final int MIN_FIRST_NAME_LENGTH = 4;
    private static final int MIN_LAST_NAME_LENGTH = 3;
    private String firstName;
    private String lastName;

    public Human(String firstName, String lastName) {
        this.setFirstName(firstName);
        this.setLastName(lastName);
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        if (Character.isLowerCase(firstName.charAt(0))){
            throw new IllegalArgumentException(
                    "Expected upper case letter!Argument: firstName");
        }

        if (firstName.length() < MIN_FIRST_NAME_LENGTH){
            throw new IllegalArgumentException(String.format(
                    "Expected length at least %d symbols!Argument: firstName",
                    MIN_FIRST_NAME_LENGTH));
        }

        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        if (Character.isLowerCase(lastName.charAt(0))){
            throw new IllegalArgumentException(
                    "Expected upper case letter!Argument: lastName");
        }

        if (lastName.length() < MIN_LAST_NAME_LENGTH){
            throw new IllegalArgumentException(String.format(
                    "Expected length at least %d symbols!Argument: lastName",
                    MIN_LAST_NAME_LENGTH));
        }

        this.lastName = lastName;
    }

    @Override
    public String toString() {
        return String.format("First Name: %s%n", this.getFirstName()) +
                String.format("Last Name: %s%n", this.getLastName());
    }
}
